<launch>

    <!-- MASTER NODE! -->
    <node name="ros_test_node" pkg="linkalibr" type="ros_test_node" output="screen" clear_params="true" required="true">

        <!-- bag topics -->
        <!-- ## Ouster, Vectornav ## -->
        <!-- <param name="topic_imu"    type="string" value="/vectornav/IMU" /> --> 
        <!-- <param name="topic_lidar"  type="string" value="/os_cloud_node/points" />        -->
        
        <!-- ## VLP16, Microstrain IMU ## -->
        <param name="topic_imu"    type="string" value="/imu/data" />
        <param name="topic_lidar"    type="string" value="/velodyne_points" />

        <!-- bag parameters -->
        <param name="path_bag"    type="string" value="/home/coui/catkin_ws/bags/0531_pos2.bag" /> 
        <param name="bag_start"   type="double" value="0.0" />
        <param name="bag_durr"    type="int"    value="-1" />

        <!-- Estimator Parameters -->
        <param name="use_fej" type="bool" value="false" />
        <param name="use_imuavg" type="bool" value="true" />
        <param name="use_rk4int" type="bool" value="true" />
        <param name="calib_lidar_timeoffset" type="bool" value="false" />
        <param name="calib_extrinsics"   type="bool"   value="true" />
        <param name="max_clones" type="int" value="2" />
        <param name="updater_do_chi2_check" type="bool" value="false"/>

        <!-- Filter Init Parameters: imu의 init을 위한 파라미터. EKF에는 영향x -->
        <param name="init_window_time" type="double" value="1"/> <!-- in second -->
        <param name="init_imu_thresh" type="double" value="0.25"/> <!-- jerk threshold -->

        <!-- IMU Noise Parameters: Datesheet 참조 -->

        <!-- ##Vectornav VN-300##  -->
        <!-- <param name="accelerometer_noise_density"  type="double"   value="0.0015825749257056074" />
        <param name="accelerometer_random_walk"    type="double"   value="4.1557212532459854e-05" />
        <param name="gyroscope_noise_density"      type="double"   value="8.97991030686202e-05" />
        <param name="gyroscope_random_walk"        type="double"   value="7.253101866424753e-07" /> -->

        <!-- Microstrain 3DM-GX5-25  -->
        <param name="accelerometer_noise_density"  type="double"   value="3.9939570888238808e-03" />
        <param name="accelerometer_random_walk"    type="double"   value="6.4356659353532566e-05" />
        <param name="gyroscope_noise_density"      type="double"   value="1.5636343949698187e-03" />
        <param name="gyroscope_random_walk"        type="double"   value="3.5640318696367613e-05" />




        <!-- Updater Parameters -->
        <!-- Chi2 Parameters -->
        <param name="updater_chi2_multiplier" type="int" value="1" />
        <!-- Update Noise (sigma) Params -->
        <param name="updater_rotation_noise" type="double" value="0.1" />
        <param name="updater_translation_noise" type="double" value="0.1" />

        <!-- State Initial Noise (sigma) Params -->
        <param name="state_init_x_noise" type="double" value="0.5" />
        <param name="state_init_y_noise" type="double" value="0.5" />
        <param name="state_init_z_noise" type="double" value="0.5" />
        <param name="state_init_rx_noise" type="double" value="0.1" />
        <param name="state_init_ry_noise" type="double" value="0.1" />
        <param name="state_init_rz_noise" type="double" value="0.1" />

        <param name="state_init_timeoffset_noise" type="double" value="0.01" />

        <!-- IMU Gravity -->

        <!-- ##Vectornav VN-300##  -->
        <!-- <rosparam param="gravity">[0.0, 0.0, 9.81]</rosparam> -->

        <!-- Microstrain 3DM-GX5-25  -->
        <rosparam param="gravity">[0.0, 0.0, 9.80511]</rosparam>
        

        <!-- initialization of lidar imu time offset -->
        <param name="calib_lidarimu_dt" type="double" value="0.0" />

        <!-- params determining map size -->
        <param name="limit_map_size" type="bool" value="false"/>
        <!-- no_of_scans 는 바꿔도 차이 없음 -->
        <param name="no_of_scans_for_map" type="int" value="400"/>     
        

        <!-- params for downsampling clouds before mapping -->
        <param name="downsample_for_mapping" type="bool" value="false"/>

        <!-- initial calibration file -->
        <param name="init_calibration_result_filename" type="string" value="$(find linkalibr)/data/coui/I_T_L_init.txt"/>

        <!-- output filenames -->
        <param name="lin_trajectory_filename" type="string" value="/home/coui/catkin_ws/src/imu_lidar_calibration/linkalibr/data/coui/lin_trajectory_undistortedcloud.csv"/>
        <param name="lin_bias_filename" type="string" value="/home/coui/catkin_ws/src/imu_lidar_calibration/linkalibr/data/coui/lin_bias_undistortedcloud.csv"/>
        <param name="lin_velocity_filename" type="string" value="/home/coui/catkin_ws/src/imu_lidar_calibration/linkalibr/data/coui/lin_velocity_undistortedcloud.csv"/>
        <param name="lin_calib_extrinsic_filename" type="string" value="/home/coui/catkin_ws/src/imu_lidar_calibration/linkalibr/data/coui/lin_calib_extrinsics_undistortedcloud.csv"/>
        <param name="lin_calib_dt_filename" type="string" value="/home/coui/catkin_ws/src/imu_lidar_calibration/linkalibr/data/coui/lin_calib_dt_undistortedcloud.csv"/>

        <param name="lo_trajectory_filename" type="string" value="/home/coui/catkin_ws/src/imu_lidar_calibration/linkalibr/data/coui/lo_trajectorylin_trajectory_undistortedcloud.csv"/>

        <!-- NDT Resolution -->
        <param name="ndt_resolution" type="double" value="0.1"/>

        <!-- Undistortion flag-->
        <param name="do_undistortion" type="bool" value="true" />

        <!-- Do planar update -->
        <param name="do_planar_update" type="bool" value="false"/>

        <!-- Write an output -->
        <param name="calibration_result_filename" type="string" value="/home/coui/catkin_ws/src/imu_lidar_calibration/linkalibr/data/coui/I_T_L_final.txt"/>
        <param name="gen_map_data" type="bool" value="true"/>
        <param name="map_csv_file_name" type="string" value="/home/coui/catkin_ws/src/imu_lidar_calibration/linkalibr/data/coui/map.csv"/>

       <!-- location to store cloud -->
       <!-- <param name="raw_scan_folder_name" type="string" value="/home/coui/catkin_ws/src/linkalibr/scans/raw"/>
       <param name="deskewed_scan_folder_name" type="string" value="/home/coui/catkin_ws/src/linkalibr/scans/deskewed"/> -->

       <param name="plane_params_filename" type="string" value="/home/coui/catkin_ws/src/linkalibr/data/coui/forOptimization/plane_params.csv"/>
       <param name="planar_points_raw_scans_filename" type="string" value="/home/coui/catkin_ws/src/linkalibr/data/coui/forOptimization/planar_points_raw_scans.csv"/>
       <param name="planar_points_deskewed_map_filename" type="string" value="/home/coui/catkin_ws/src/linkalibr/data/coui/forOptimization/planar_points_deskewed_map.csv"/>
       <param name="surfel_map_filename" type="string" value="/home/coui/catkin_ws/src/linkalibr/data/coui/forOptimization/surfel_map.csv"/>

       <param name="imu_measurements_filename" type="string" value="/home/coui/catkin_ws/src/linkalibr/data/coui/forOptimization/imu_measurements.csv"/>
       <param name="imu_trajectory_filename" type="string" value="/home/coui/catkin_ws/src/linkalibr/data/coui/forOptimization/imu_trajectory.csv"/>
       <param name="lidar_trajectory_filename" type="string" value="/home/coui/catkin_ws/src/linkalibr/data/coui/forOptimization/lidar_trajectory.csv"/>

    </node>

</launch>